### PR TITLE
Propagate the reason for Eviction into the pod TerminationTarget condition

### PR DESCRIPTION
**Pull request by pajakd**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
Whenever kueue is stopping a pod, it adds a condition TerminationTarget. However, the reason for termination is always "StoppedByKueue". The goal of this PR is to set a specific reason depending on why the pod is stopped (like workload is evicted or workload is deleted).

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
The reason for stopping a pod is now specified in the pod TerminationTarget condition
```

### CODE REVIEW

File: pkg/controller/jobframework/interface.go [hunk 1]
```diff
@@ -62,13 +62,13 @@ type JobWithReclaimablePods interface {
...
-type StopReason int
+type StopReason string
 
 const (
-	StopReasonWorkloadDeleted StopReason = iota
-	StopReasonWorkloadEvicted
-	StopReasonNoMatchingWorkload
-	StopReasonNotAdmitted
+	StopReasonWorkloadDeleted    StopReason = "WorkloadDeleted"
+	StopReasonWorkloadEvicted    StopReason = "WorkloadEvicted"
```
Comment (alculquicondor) [inline pkg/controller/jobframework/interface.go:69]: Uhm... this one might not be enough. We also need to know if it was WaitForPodsReady or Preemption

Comment (pajakd) [reply pkg/controller/jobframework/interface.go:69]: Ok, a second try. Please take a look. I don't know if I should add to this enum all possible Eviction reasons or I can just directly cast and propagate the `evCond.Reason` reason?

Comment (alculquicondor) [reply pkg/controller/jobframework/interface.go:69]: As discussed offline, let's see if we can use a hyphen to concatenate the reasons.

Comment (tenzen-y) [reply pkg/controller/jobframework/interface.go:69]: @alculquicondor Could we decide the delimiter based on this discussion https://github.com/kubernetes/enhancements/pull/4479/files#r1581226166 since the reason should be CamelCase?

> In condition types, and everywhere else they appear in the API, Reason is intended to be a one-word, CamelCase representation of the category of cause of the current status

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties

That discussion will seem to happen at the next sig apps meeting.

Comment (tenzen-y) [reply pkg/controller/jobframework/interface.go:69]: > That discussion will seem to happen at the next sig apps meeting.

Oh, there is not the discussion in the agenda...

Comment (alculquicondor) [reply pkg/controller/jobframework/interface.go:69]: I'm just trying to make the reason make sense.

`WorkloadEvictedPreempted` doesn't read that well.

So, the alternatives would be:
1. `WorkloadPreempted` (drop the `Evicted` part)
2. `Preempted` (just use the reason directly, and forget about the Workload bit)
3. `WorkloadEvictedBecausePreempted` (same proposal as the hyphen, but with a word)

Maybe 1 reads best of all? WDYT?

Comment (tenzen-y) [reply pkg/controller/jobframework/interface.go:69]: I agree with you since we have already the `Evicted` in the `condition.type`, right?

File: pkg/controller/jobframework/reconciler.go [hunk 2]
```diff
@@ -761,7 +761,13 @@ func (r *JobReconciler) stopJob(ctx context.Context, job GenericJob, wl *kueue.W
 	}
 
 	if jws, implements := job.(ComposableJob); implements {
-		stoppedNow, err := jws.Stop(ctx, r.client, info, stopReason, eventMsg)
+		reason := stopReason
```
Comment (trasc) [inline pkg/controller/jobframework/reconciler.go:764]: In my opinion, ignoring the received reason just for ComposableJob when Evicted condition is hard to follow, 
the caller `stopJob` should provide the right reason for all the jobs.

File: pkg/controller/jobframework/interface.go (hunk 1 above)
Comment (pajakd) [reply pkg/controller/jobframework/interface.go:70]: So the possible reasons for eviction are: https://github.com/kubernetes-sigs/kueue/blob/b870144dc272788a755369411052bf5b81e50348/apis/kueue/v1beta1/workload_types.go#L332-L351

And 1. reads definitely good for "Preempted" but for the other ones it would be a bit confusing. In my opinion concatenation with the dash would make the most sense. But if it is not recommended in the guidelines then we can decide on any of the alternatives. Would `WorkloadEvictedWithReason{Preempted,PodsReadyTimeout ...}` (a variant of 3) be too long?

Comment (alculquicondor) [reply pkg/controller/jobframework/interface.go:70]: I like it, even if it's too long

File: pkg/controller/jobframework/reconciler.go (hunk 2 above)
Comment (alculquicondor) [reply pkg/controller/jobframework/reconciler.go:823]: I agree, but right now we don't have a condition in every job before suspending.

We could work on that as a follow up. For now, the pain point is in the pod integration, given that they need to be replaced and it's not now immediately obvious when a pod was deleted because of an Eviction (because the Workload condition can change if it's immediately re-admitted).

File: pkg/controller/jobframework/interface.go (hunk 1 above)
Comment (pajakd) [reply pkg/controller/jobframework/interface.go:70]: Ok, changed the joining word to "Reason".

File: pkg/controller/jobframework/reconciler.go [hunk 3]
```diff
@@ -820,7 +820,13 @@ func (r *JobReconciler) stopJob(ctx context.Context, job GenericJob, wl *kueue.W
...
+		if stopReason == StopReasonWorkloadEvicted {
+			if evCond := apimeta.FindStatusCondition(wl.Status.Conditions, kueue.WorkloadEvicted); evCond != nil && evCond.Status == metav1.ConditionTrue {
```
Comment (mimowo) [inline pkg/controller/jobframework/reconciler.go:825]: You can use `apimeta.IsStatusConditionTrue`

File: pkg/controller/jobframework/reconciler.go [hunk 4]
```diff
@@ -820,7 +820,13 @@ func (r *JobReconciler) stopJob(ctx context.Context, job GenericJob, wl *kueue.W
...
+				reason = StopReason(string(StopReasonWorkloadEvicted) + "Reason" + evCond.Reason)
```
Comment (mimowo) [inline pkg/controller/jobframework/reconciler.go:826]:
```suggestion
				reason = StopReason(string(StopReasonWorkloadEvicted) + "DueTo" + evCond.Reason)
```
Maybe it reads better?

Comment (pajakd) [reply pkg/controller/jobframework/reconciler.go]: Ok. It definitely reads better.

File: pkg/controller/jobframework/reconciler.go (hunk 3 above)
Comment (pajakd) [reply pkg/controller/jobframework/reconciler.go:825]: Right, but I do need the entire `Condition` object (not only its presence/absence) because I have to read the `Reason` from it.

Comment (mimowo) [reply pkg/controller/jobframework/reconciler.go:825]: oh, yes

**Review by alculquicondor — COMMENTED** [review]
/lgtm
/approve

**Comment by alculquicondor** [comment]
/release-note-edit

```release-note
Pod Integration: The reason for stopping a pod is now specified in the pod `TerminationTarget` condition
```
